add documentation for using an external event manager (#1944)
private-octopus/picoquic: doc/howto_use_external_event_manager.md

QUIC correctness depends on accurate PTO (Probe Timeout) timers. Picoquic exposes `picoquic_get_next_wake_delay`, which returns how long the stack can sleep before it must run again (`picoquic/quicctx.c:1433`). A typical integration stores a periodic timer event in the loop and re-arms it after every send/receive:

```c
static void on_timeout(evutil_socket_t fd, short what, void *arg) {
    struct quic_loop_ctx *ctx = arg;
    quic_schedule_send(ctx); /* Ensure PTO probes are emitted */
    quic_reschedule_timer(ctx);
}

static void quic_reschedule_timer(struct quic_loop_ctx *ctx) {
    const int64_t delay_max_us = 1000 * 1000; /* clamp to 1 second */
    uint64_t now = picoquic_current_time();
    int64_t delay = picoquic_get_next_wake_delay(ctx->quic, now, delay_max_us);
    struct timeval tv = {
        .tv_sec = (time_t)(delay / 1000000),
        .tv_usec = (suseconds_t)(delay % 1000000)
    };
    evtimer_add(ctx->timer_event, &tv);
}
```

Call `quic_reschedule_timer` once after creating the QUIC context and again after every send or receive so the timer adapts to new wake-ups.

## 4. Producing Packets for Transmission

> Whenever the UDP socket becomes writable (or immediately after processing inbound packets), call `picoquic_prepare_next_packet_ex`. The function returns either a stateless packet (retry/close) or the next connection’s datagram (`picoquic/sender.c:4227`). You pass in buffers and Picoquic fills in the destination address and interface index.

Reviewer: ```suggestion
Whenever the UDP socket becomes writable (or immediately after processing inbound packets),
call `picoquic_prepare_next_packet_ex`.
The function returns either a stateless packet (retry/close) or the next connection’s datagram
(`picoquic/sender.c:4227`). You pass in buffers and Picoquic fills in the destination address and interface index.
In some case, the function will return nothing (send_length == 0) -- the application just needs to update its
timers, as explained in the "scheduling timers" section.
```

    for (;;) {
        int ret = picoquic_prepare_next_packet_ex(ctx->quic, picoquic_current_time(),
            send_buffer, sizeof(send_buffer), &send_length,
            &addr_to, &addr_from, &if_index, &log_cid, &last_cnx, NULL);

        if (ret != 0 || send_length == 0) {
            break; /* Nothing to send now */
        }

        ssize_t nb = sendto(ctx->udp_fd, send_buffer, send_length, 0,
            (struct sockaddr*)&addr_to, (socklen_t)picoquic_addr_length((struct sockaddr*)&addr_to));
        if (nb < 0) {
            if (errno == EAGAIN || errno == EWOULDBLOCK) {
                /* Re-arm writable notification and retry later. */
                break;
            }
            perror("sendto");
            break;
        }
    }

    quic_reschedule_timer(ctx);
}
```

After each iteration, continue polling until `send_length` becomes zero so that stateless packets and retransmissions are flushed immediately. The helper `picoquic_addr_length` computes the correct `socklen_t` for IPv4 or IPv6 destinations, matching the utility used throughout the built-in socket loop (`picoquic/util.c:541`).

### Handling Multiple Paths or Alternate Ports

> If your application binds multiple sockets (for multipath or alternate ports), call `picoquic_prepare_next_packet_ex` once per writable socket and respect the returned `addr_from` / interface hints. Packet production is coordinated with each path’s congestion controller during `picoquic_prepare_packet_ex`, which chooses the path tuple providing the datagram (`picoquic/sender.c:4055`).

Reviewer: ```suggestion
If your application binds multiple sockets (for multipath or alternate ports), call `picoquic_prepare_next_packet_ex` and send the packet that was produced on
the socket corresponding to the returned `addr_from` / interface. The code will determine the outgoing interface and source address based on the internal path selection logic. It does not monitor the "available for sending" API per socket; instead, it relies on per-path congestion control to ensure that each interface only gets a reasonable amount of traffic.
```

### ECN and Ancillary Data

`picoquic_incoming_packet_ex` accepts an ECN codepoint so that congestion controllers such as L4S or Prague can react to marks (`picoquic/packet.c:2376`). If your network stack surfaces ECN values, translate them into the `received_ecn` byte before calling Picoquic. Similarly, you can decode interface indices or destination addresses from `recvmsg` control data and fill the `addr_to`/`if_index` arguments.

### Stateless Packets and Logging

Because `picoquic_prepare_next_packet_ex` checks the stateless packet queue first (`picoquic/sender.c:4233`), your integration automatically handles Retry, Version Negotiation, and close frames even before a connection is fully created. Ensure your event loop sends every datagram returned by the API—even during connection setup or DoS mitigation.

### Packet Coalescing and Maximum MTU

For best throughput, keep the send buffer at least `PICOQUIC_MAX_PACKET_SIZE` bytes. Picoquic will coalesce multiple QUIC packets into a single UDP datagram when GSO is available (`picoquic/sockloop.c:1006`), but falling back to standard `sendto` still works—the function simply returns the exact length to transmit.

> 

Reviewer: Not quite. There are two separate concepts, coalescing and UDP GSO.

Coalescing is used internally, to format the content of a UDP payload. It does require that the buffer be at least 1200 bytes (the minimum length of a handshake packet), but it will work fine with larger buffers.

UDP GSO allows picoquic to prepare several UDP packets on a single call. All these packets will have the "addr_from", "addr_to" and "interface" parameters. All but the last one will have the same packet size. To use UDP GSO, the application should provide a buffer capable of holding the payload of multiple packets -- a size of at least 16K is nice, the current code uses 64K. UDP GSO is controlled by the parameter `send_msg_size` in:
~~~
int picoquic_prepare_packet_ex(picoquic_cnx_t* cnx,
    uint64_t current_time, uint8_t* send_buffer, size_t send_buffer_max, size_t* send_length,
    struct sockaddr_storage* p_addr_to, struct sockaddr_storage* p_addr_from, int* if_index,
    size_t* send_msg_size);
~~~
If that parameter is a NULL pointer, UDP GSO is not used. Otherwise, the code will attempt to format multiple packets in the `send_buffer`. The `send_length` will return the combined length of all these packets, and the `send_msg_size` will return the length of the first packet. The packets after the first will have the same length, except for the last one which may be shorter, and would carry the remaining bytes in the `send_buffer`.

### Error Handling and Connection Cleanup

If `picoquic_prepare_next_packet_ex` returns an error, inspect the connection (`last_cnx`) and call `picoquic_close` or `picoquic_free` as needed. Connection states progress through `picoquic_state_client_ready`, `picoquic_state_disconnected`, etc., just as they do in the built-in loop (`sample/sample_client.c:380`).

> 

Reviewer: I don't think that's necessary. The event management does not have to concern itself with connection handling. If a connection fails, the application will receive a callback such as:
~~~
    picoquic_callback_stateless_reset, /* Stateless reset received from peer. Stream=0, bytes=NULL, len=0 */
    picoquic_callback_close, /* Connection close. Stream=0, bytes=NULL, len=0 */
    picoquic_callback_application_close, /* Application closed by peer. Stream=0, bytes=NULL, len=0 */
~~~
The exception is if we have an internal error, indicating that picoquic has lost its internal state and must terminate.
In that case, calling `picoquic_free` to free the `quic` handle is appropriate.

And yes, the builtin loop has a bit of complicated logic here. That's because it is also used for testing and debugging, and applications do not normally have to do that.

## 7. Checklist

1. Instantiate the QUIC context once at startup.
2. Register UDP sockets with your event manager and pass received datagrams into `picoquic_incoming_packet_ex`.
3. Poll `picoquic_prepare_next_packet_ex` after every receive or wake-up and send any returned datagrams immediately.
4. Reschedule a timer using `picoquic_get_next_wake_delay` so PTOs trigger correctly.

> 5. Use Picoquic’s logging (`picoquic_log_packet`) or your own instrumentation to monitor traffic.

Reviewer: ```suggestion
5. Free the QUIC context using `void picoquic_free(picoquic_quic_t* quic)` when closing the event loop. This will
    trigger the appropriate "connection close" callbacks, allowing the application to properly close its connection
    contexts and ensuring that log files are properly flushed and closed.
```

> 

Reviewer: The application can use one of the logging providers (text, binary or qlog) to get a log. It should not have to `log packet` from within the event manager. But consider "5. Free the QUIC context once when closing the event loop." That will trigger the appropriate "connection close" callbacks, allowing the application to properly close its connection contexts.
